// beastsaber-api/lib/AxiosRequestHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AxiosRequestHandler = void 0;
const tslib_1 = require("tslib");
const axios_1 = (0, tslib_1.__importDefault)(require("axios"));
const axios_cookiejar_support_1 = require("axios-cookiejar-support");
const tough_cookie_1 = require("tough-cookie");
class AxiosRequestHandler {
    constructor(domain, cookieStore) {
        this.domain = domain;
        this.cookieJar = new tough_cookie_1.CookieJar(cookieStore);
        this.axios = (0, axios_cookiejar_support_1.wrapper)(axios_1.default.create({
            jar: this.cookieJar,
        }));
    }
    get({ endpoint, headers, }) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const url = `https://${this.domain}${endpoint}`;
            const config = {
                headers: headers,
            };
            const response = yield this.axios.get(url, config);
            return {
                url: url,
                headers: response.headers,
                body: response.data,
                status: response.status,
            };
        });
    }
    post({ endpoint, headers, params, }) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const url = `https://${this.domain}${endpoint}`;
            const data = new URLSearchParams(params);
            const config = {
                headers: headers,
                maxRedirects: 0,
                validateStatus: (status) => status >= 200 && status < 400,
            };
            const response = yield this.axios.post(url, data, config);
            return {
                url: url,
                headers: response.headers,
                body: response.data,
                status: response.status,
            };
        });
    }
}
exports.AxiosRequestHandler = AxiosRequestHandler;
