// beastsaber-api/lib/BeastSaber.js
"use strict";
const tslib_1 = require("tslib");
const tough_cookie_1 = require("tough-cookie");
const AxiosRequestHandler_1 = require("./AxiosRequestHandler");
const BeastSaberEndpoints_1 = require("./BeastSaberEndpoints");
const LoginFailedError_1 = (0, tslib_1.__importDefault)(require("./errors/LoginFailedError"));
const html_map_list_parser_1 = require("./parser/html-map-list-parser");
const html_map_single_parser_1 = require("./parser/html-map-single-parser");
class BeastSaber {
    /**
     * @param cookieStore a store for a CookieJar to use with the default request handler
     */
    constructor(domain, cookieStore) {
        this.userAgent = "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:95.0) Gecko/20100101 Firefox/95.0";
        this.domain = domain !== null && domain !== void 0 ? domain : "bsaber.com";
        this.handler = new AxiosRequestHandler_1.AxiosRequestHandler(this.domain, cookieStore);
        this.baseUrl = "https://" + this.domain;
    }
    /**
     * Set a custom request handler.
     */
    setRequestHandler(handler) {
        this.handler = handler;
    }
    getHeaders() {
        return {
            Accept: "*/*",
            "Accept-Language": "en-US",
            Referer: "https://bsaber.com/",
            Origin: "https://bsaber.com/",
            DNT: "1",
            Pragma: "no-cache",
            "Cache-Control": "no-cache",
        };
    }
    /**
     * Login the user.
     * @param userLogin user login/email
     * @param password user password
     * @returns array of currently stored cookies for bsaber.com
     */
    login(userLogin, password) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const endpoint = BeastSaberEndpoints_1.BeastSaberEndpoints.Login();
            const params = {
                log: userLogin,
                pwd: password,
                rememberme: "forever",
                "wp-submit": "Log In",
                redirect_to: "https://bsaber.com/",
                testcookie: "1",
            };
            const cookie = new tough_cookie_1.Cookie({
                key: "wordpress_test_cookie",
                value: "WP+Cookie+check",
                domain: this.domain,
                path: "/",
            });
            this.handler.cookieJar.setCookie(cookie, this.baseUrl);
            const response = yield this.handler.post({
                endpoint,
                params,
                headers: this.getHeaders(),
            });
            if (response.status !== 302) {
                throw new LoginFailedError_1.default(response.body);
            }
            return yield this.handler.cookieJar.getCookies(this.baseUrl);
        });
    }
    /**
     * Check if the user is logged in (based on cookies expiration).
     */
    isLoggedIn(endpoint) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const url = endpoint ? this.baseUrl + endpoint : this.baseUrl;
            const cookies = yield this.handler.cookieJar.getCookies(url);
            return cookies.some((cookie) => cookie.TTL() > 0 && cookie.key.includes("wordpress_logged_in"));
        });
    }
    /**
     * Try to bookmark a map.
     * @param mapId bsaber.com post ID
     * @returns true if succeeded, false if failed (not logged in, etc.)
     */
    bookmarkAdd(mapId) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const endpoint = BeastSaberEndpoints_1.BeastSaberEndpoints.AdminAjax();
            if (!this.isLoggedIn(endpoint))
                return false;
            const params = {
                action: "bsaber_bookmark_post",
                type: "add_bookmark",
                post_id: String(mapId),
            };
            const response = yield this.handler.post({
                endpoint,
                params,
                headers: this.getHeaders(),
            });
            return response.body.length === 0;
        });
    }
    /**
     * Try to un-bookmark a map.
     * @param mapId bsaber.com post ID
     * @returns true if succeeded, false if failed (not logged in, etc.)
     */
    bookmarkRemove(mapId) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const endpoint = BeastSaberEndpoints_1.BeastSaberEndpoints.AdminAjax();
            if (!this.isLoggedIn(endpoint))
                return false;
            const params = {
                action: "bsaber_bookmark_post",
                type: "remove_bookmark",
                post_id: String(mapId),
            };
            const response = yield this.handler.post({
                endpoint,
                params,
                headers: this.getHeaders(),
            });
            return response.body.length === 0;
        });
    }
    /**
     * Get a single map.
     * @param key map hexadeximal key
     * @returns Map or null if parse error
     */
    getMapByKey(key) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const response = yield this.handler.get({
                endpoint: BeastSaberEndpoints_1.BeastSaberEndpoints.MapByKey(key),
            });
            return yield (0, html_map_single_parser_1.parseMapSingle)(response.body, key);
        });
    }
    /**
     * Get a single map.
     * @param id bsaber.com post ID
     * @returns Map or null if parse error
     */
    getMapById(id) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const response = yield this.handler.get({
                endpoint: BeastSaberEndpoints_1.BeastSaberEndpoints.MapById(id),
            });
            return yield (0, html_map_single_parser_1.parseMapSingle)(response.body, id);
        });
    }
    /**
     * Fetch maps by the given params.
     */
    getMaps(params) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const response = yield this.handler.get({
                endpoint: BeastSaberEndpoints_1.BeastSaberEndpoints.Songs(params),
            });
            return yield (0, html_map_list_parser_1.parseMapList)(response.body);
        });
    }
    /**
     * Get maps in a playlist.
     * @param playlistSlug playlist slug name, as seen in the page URL
     */
    getPlaylistMaps(playlistSlug, page = 1) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const response = yield this.handler.get({
                endpoint: BeastSaberEndpoints_1.BeastSaberEndpoints.Playlist(playlistSlug, page),
            });
            return yield (0, html_map_list_parser_1.parseMapList)(response.body);
        });
    }
    /**
     * Get maps uploaded by a user.
     */
    getUploadedBy(userLogin, page = 1) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            return yield this.getMaps({
                uploadedBy: userLogin,
                page: page,
                sortOrder: "new",
            });
        });
    }
    /**
     * Get maps bookmarked by a user.
     */
    getBookmarkedBy(userLogin, page = 1) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            return yield this.getMaps({
                bookmarkedBy: userLogin,
                page: page,
                sortOrder: "new",
            });
        });
    }
    /**
     * Get maps by mappers followed by a user.
     */
    getFollowedBy(userLogin, page = 1) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            return yield this.getMaps({
                followedBy: userLogin,
                page: page,
                sortOrder: "new",
            });
        });
    }
    /**
     * Get most reviewed maps.
     */
    getMostReviewed(page = 1) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const response = yield this.handler.get({
                endpoint: BeastSaberEndpoints_1.BeastSaberEndpoints.MostReviewed(page),
            });
            return yield (0, html_map_list_parser_1.parseMapList)(response.body);
        });
    }
    /**
     * Get "Top User Reviews All" maps.
     */
    getTopUserReviewsAll(page = 1) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
            const response = yield this.handler.get({
                endpoint: BeastSaberEndpoints_1.BeastSaberEndpoints.TopUserReviewsAll(page),
            });
            return yield (0, html_map_list_parser_1.parseMapList)(response.body);
        });
    }
}
module.exports = BeastSaber;
