// beastsaber-api/lib/parser/html-map-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseMapElements = void 0;
const category_parser_1 = require("./category-parser");
const difficulty_parser_1 = require("./difficulty-parser");
function parseMapElements(title, url, imageUrl, root, meta) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    const id = (_a = root.querySelector("a.js-bookmark")) === null || _a === void 0 ? void 0 : _a.getAttribute("data-id");
    const difficulties = root.querySelectorAll("a.post-difficulty");
    if (!difficulties || !id)
        return null;
    const datetime = (_b = meta.querySelector("time")) === null || _b === void 0 ? void 0 : _b.getAttribute("datetime");
    const categories = meta.querySelectorAll(".bsaber-categories a");
    if (!datetime || !categories)
        return null;
    const mapperImg = meta.querySelector("img.avatar");
    const mapperSpan = meta.querySelector("span.mapper_id");
    const mapperLink = meta.querySelector("strong.mapper_id a");
    const mapperName = (_d = (_c = mapperLink === null || mapperLink === void 0 ? void 0 : mapperLink.text) === null || _c === void 0 ? void 0 : _c.trim()) !== null && _d !== void 0 ? _d : (_e = mapperSpan === null || mapperSpan === void 0 ? void 0 : mapperSpan.text) === null || _e === void 0 ? void 0 : _e.trim();
    const mapperUrl = mapperLink === null || mapperLink === void 0 ? void 0 : mapperLink.getAttribute("href");
    let mapper;
    if (mapperImg && mapperName && mapperUrl) {
        mapper = {
            name: mapperName,
            login: mapperUrl.split("/")[4],
            avatarUrl: mapperImg.getAttribute("src"),
            verified: !!meta.querySelector("i.yz-account-verified"),
        };
    }
    else if (mapperName) {
        mapper = {
            name: mapperName,
            login: mapperName,
        };
    }
    else {
        return null;
    }
    const upvotes = (_h = (_g = (_f = root.querySelector(".post-row span.post-stat .fa-thumbs-up")) === null || _f === void 0 ? void 0 : _f.parentNode) === null || _g === void 0 ? void 0 : _g.text) !== null && _h !== void 0 ? _h : "0";
    const downvotes = (_l = (_k = (_j = root.querySelector(".post-row span.post-stat .fa-thumbs-down")) === null || _j === void 0 ? void 0 : _j.parentNode) === null || _k === void 0 ? void 0 : _k.text) !== null && _l !== void 0 ? _l : "0";
    const listenOnClick = (_m = root
        .querySelector("a.js-listen")) === null || _m === void 0 ? void 0 : _m.getAttribute("onclick");
    const hash = listenOnClick === null || listenOnClick === void 0 ? void 0 : listenOnClick.match(/previewSong\(this, 'https:\/\/cdn\.beatsaver\.com\/([a-f0-9]+)\./);
    if (!hash)
        return null;
    return {
        id: parseInt(id, 10),
        key: url.split("/")[4],
        hash: hash[1],
        title: title,
        imageUrl: imageUrl,
        datePublished: new Date(datetime),
        tags: (0, category_parser_1.parseCategories)(categories),
        difficulties: (0, difficulty_parser_1.parseDifficulties)(difficulties),
        mapper: mapper,
        curator: undefined,
        upvotes: parseInt(upvotes, 10),
        downvotes: parseInt(downvotes, 10),
        excerpt: undefined,
        description: undefined,
    };
}
exports.parseMapElements = parseMapElements;
