// beastsaber-api/lib/parser/html-map-list-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseMapList = void 0;
const tslib_1 = require("tslib");
const node_html_parser_1 = (0, tslib_1.__importDefault)(require("node-html-parser"));
const html_map_parser_1 = require("./html-map-parser");
function parseMapList(html) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
        const root = (0, node_html_parser_1.default)(html);
        const result = {
            maps: [],
            pagination: {
                current: 1,
            },
        };
        root.querySelectorAll("article.type-post").forEach((article) => {
            var _a, _b, _c, _d, _e;
            const postGallery = article.querySelector(".post-gallery");
            const galleryLink = postGallery === null || postGallery === void 0 ? void 0 : postGallery.querySelector("a[rel=bookmark]");
            const title = galleryLink === null || galleryLink === void 0 ? void 0 : galleryLink.getAttribute("title");
            const url = galleryLink === null || galleryLink === void 0 ? void 0 : galleryLink.getAttribute("href");
            const imageUrl = (_a = galleryLink === null || galleryLink === void 0 ? void 0 : galleryLink.querySelector("img")) === null || _a === void 0 ? void 0 : _a.getAttribute("data-original");
            if (!title || !url || !imageUrl)
                return;
            const meta = article.querySelector("div.post-mapper-id-meta");
            if (!meta)
                return;
            const map = (0, html_map_parser_1.parseMapElements)(title, url, imageUrl, article, meta);
            if (!map)
                return;
            const curatorLink = article.querySelector("div.post-row > span > a[rel=author]");
            const curatorName = (_b = curatorLink === null || curatorLink === void 0 ? void 0 : curatorLink.text) === null || _b === void 0 ? void 0 : _b.trim();
            const curatorUrl = curatorLink === null || curatorLink === void 0 ? void 0 : curatorLink.getAttribute("href");
            let curator;
            if (curatorName && curatorUrl) {
                curator = {
                    name: curatorName,
                    login: curatorUrl.split("/")[4],
                };
            }
            map.excerpt = (_e = (_d = (_c = article
                .querySelector(".post-content")) === null || _c === void 0 ? void 0 : _c.text) === null || _d === void 0 ? void 0 : _d.replace(`Mapper: ${map.mapper.name}`, "")) === null || _e === void 0 ? void 0 : _e.trim();
            map.curator = curator;
            result.maps.push(map);
        });
        if (result.maps) {
            result.pagination.current = 1;
            result.pagination.max = 1;
        }
        const pageCurrent = root.querySelector(".page-numbers-current");
        if (pageCurrent) {
            result.pagination.current = parseInt(pageCurrent.text.trim(), 10);
        }
        const pagePrevious = root.querySelector(".page-numbers.prev");
        if (pagePrevious) {
            result.pagination.previous = result.pagination.current - 1;
        }
        const pageNext = root.querySelector(".page-numbers.next");
        if (pageNext) {
            result.pagination.next = result.pagination.current + 1;
        }
        const pageMax = root.querySelectorAll(".page-numbers:not(.next)");
        if (pageMax.length) {
            result.pagination.max = parseInt(pageMax[pageMax.length - 1].text.trim(), 10);
        }
        return result;
    });
}
exports.parseMapList = parseMapList;
